### ISSUE TITLE
Argparse and multiprocessing breaks nuitka 

### ISSUE THREAD

**Issue by Nixccc** [Nuitka/Nuitka#3018]
Python: 3.10.11 (tags/v3.10.11:7d4cc5a, Apr  5 2023, 00:38:17) [MSC v.1929 64 bit (AMD64)]
    Flavor: CPython Official
    Executable: C:\Users\G4U5AAG\Documents\prj\edd-compare\edd-compare\.venv\Scripts\python.exe
    OS: Windows
    Arch: x86_64
    WindowsRelease: 10
    Version C compiler: C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.40.33807\bin\Hostx64\x64\cl.exe (cl 14.3). (MINGW also tried)

- How did you install Nuitka and Python

Python installed via python website as standalone
Nuitka installed via pip install nuitka

```
Package            Version     Location                                                                       Installer
------------------ ----------- ------------------------------------------------------------------------------ ---------
annotated-types    0.7.0       c:\users\g4u5aag\documents\prj\edd-compare\edd-compare\.venv\lib\site-packages pip
certifi            2024.7.4    c:\users\g4u5aag\documents\prj\edd-compare\edd-compare\.venv\lib\site-packages pip
cffi               1.16.0      c:\users\g4u5aag\documents\prj\edd-compare\edd-compare\.venv\lib\site-packages pip
chardet            5.2.0       c:\users\g4u5aag\documents\prj\edd-compare\edd-compare\.venv\lib\site-packages pip
charset-normalizer 3.3.2       c:\users\g4u5aag\documents\prj\edd-compare\edd-compare\.venv\lib\site-packages pip
contourpy          1.2.1       c:\users\g4u5aag\documents\prj\edd-compare\edd-compare\.venv\lib\site-packages pip
cryptography       42.0.8      c:\users\g4u5aag\documents\prj\edd-compare\edd-compare\.venv\lib\site-packages pip
cycler             0.12.1      c:\users\g4u5aag\documents\prj\edd-compare\edd-compare\.venv\lib\site-packages pip
Nuitka             2.3.11      c:\users\g4u5aag\documents\prj\edd-compare\edd-compare\.venv\lib\site-packages pip
setuptools         65.5.0      c:\users\g4u5aag\documents\prj\edd-compare\edd-compare\.venv\lib\site-packages pip
six                1.16.0      c:\users\g4u5aag\documents\prj\edd-compare\edd-compare\.venv\lib\site-packages pip
tomli              2.0.1       c:\users\g4u5aag\documents\prj\edd-compare\edd-compare\.venv\lib\site-packages pip
typing_extensions  4.12.2      c:\users\g4u5aag\documents\prj\edd-compare\edd-compare\.venv\lib\site-packages pip
tzdata             2024.1      c:\users\g4u5aag\documents\prj\edd-compare\edd-compare\.venv\lib\site-packages pip
urllib3            2.2.2       c:\users\g4u5aag\documents\prj\edd-compare\edd-compare\.venv\lib\site-packages pip
wrapt              1.16.0      c:\users\g4u5aag\documents\prj\edd-compare\edd-compare\.venv\lib\site-packages pip
zipp               3.19.2      c:\users\g4u5aag\documents\prj\edd-compare\edd-compare\.venv\lib\site-packages pip
zstandard          0.23.0      c:\users\g4u5aag\documents\prj\edd-compare\edd-compare\.venv\lib\site-packages pip

```

```Python
from __future__ import annotations

import argparse
from multiprocessing import Process

parser = argparse.ArgumentParser(description="Compare tool GUI")
dev_args = parser.add_argument_group("development arguments (ignore if you are not a developer)")
dev_args.add_argument("--show-dev", action="store_true", help="Unlocks developer specific functionality")


def test(prgm, action):
    print(prgm, action)


def dotest(prgm, action):
    p = Process(target=test, args=(prgm, action))
    p.start()


args = parser.parse_args()
if __name__ == "__main__":
    dotest("test_compile", "starting")
```
**Result:**
```
usage: test_compile.exe [-h] [--show-dev]
test_compile.exe: error: unrecognized arguments: --multiprocessing-fork parent_pid=24356 pipe_handle=288
```

Nuitka options used:

`  python -m nuitka test_compile.py --standalone --follow-imports 
`
When adding a argparser nuitka multprocessing support breaks, searching through internet they said to add multiprocessing.freeze_support() but that didn't help in my case
https://github.com/pyinstaller/pyinstaller/wiki/Recipe-Multiprocessing

**Comment by amitie10g** [comment]
The same happened to me at GitHub Actions when using the 2.4 version.

**Comment by Nixccc** [comment]
Hmmm using 2.3.11 and it's an issue here and also in 2.3.10

**Comment by kayhayen** [comment]
There was a bug for arguments with 2.4 that has been fixed in 2.4.1 release. The multiprocessing plugin should be catching those arguments and handle it. Your original output doesn't include the Nuitka version, which is very strange.

**Comment by kayhayen** [comment]
It's in the pip listing, 2.3.11 was installed, so indeed it's unrelated. However, the 2.4 bug could have caused the symptom of multiprocessing not working too, but not in this exact form, I believe this must be an unrecognized call.

**Comment by Nixccc** [comment]
Running the program without argparse works fine for me on nuitka version 2.4.1, but the same error occurs when running with argparse

**Comment by kayhayen** [comment]
No, I think it's the way you use multiprocessing, manually with Process. That's not recognized. I am surprised that the fork-bomb handling didn't kick it, I need to check what it is. On Linux it works, but there forks do not happen, so now I need to try Windows.

**Comment by kayhayen** [comment]
So, I manage to reproduce this, the debug of self forking tell me it's executing like this

```
Commandline: 'Mini.dist\Mini.exe' 1
Commandline: '"C:\Users\kayha\repos\Py2C\Mini.dist\Mini.exe" "--multiprocessing-fork" "parent_pid=34060" "pipe_handle=300"' 4
```

For this we got special code, not sure why it doesn't kick in, probably being too picky or limited.

**Comment by kayhayen** [comment]
Ah, I think I see what's going on, so argparse looks at the arguments, before Nuitka calls the spawn function at all, so it's still wrong at the time, because standard Python cleaned it up, while for Nuitka, this will only happen after the `__parents_main__` code has run. So Nuitka injects its code at the end of the file, probably so as to not harm line numbers, but I think also `__main__` module code is expected to run and have names and imports done, not sure now.

**Comment by kayhayen** [comment]
It's also `__mp_main__` which I do not recall seeing ever, maybe for that rules are different, I need to check that out.

**Comment by Nixccc** [comment]
> So, I manage to reproduce this, the debug of self forking tell me it's executing like this
> 
> ```
> Commandline: 'Mini.dist\Mini.exe' 1
> Commandline: '"C:\Users\kayha\repos\Py2C\Mini.dist\Mini.exe" "--multiprocessing-fork" "parent_pid=34060" "pipe_handle=300"' 4
> ```
> 
> For this we got special code, not sure why it doesn't kick in, probably being too picky or limited.

Yes, exact issue that Nuitka flag prints for me aswell

**Comment by kayhayen** [comment]
I only ever saw `__parents_main__` so far. For you as a workaround, have argparse only imported below a `main()` will work better, but I will be looking into what sys.argv should be like when executing `parents_main`. It seems as if the multiprocessing code wants to post execute code and add its there, but we have better mechanisms for that, and that using the sys.argv there should be made possible in just a different way and things would work in a compatible way.

Having the multiprocessing sys.argv visible during the main re-run is not compatible. We would have to store it somehow and then get it back. I need to see if this is supposed to be the one the original program launched with or empty.

**Comment by Nixccc** [comment]
An quick update:
```Python
from __future__ import annotations

import argparse

from multiprocessing import Process, freeze_support


def test(prgm, action):
    print(prgm, action)


def dotest(prgm, action):
    p = Process(target=test, args=(prgm, action))
    p.start()


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Compare tool GUI")
    dev_args = parser.add_argument_group("development arguments (ignore if you are not a developer)")
    dev_args.add_argument("--show-dev", action="store_true", help="Unlocks developer specific functionality")
    args = parser.parse_args()
    freeze_support()
    dotest("test_compile", "starting")
```
Putting the code for argument handling in main seems to solve the issue

**Comment by kayhayen** [comment]
Well, yes, that's a workaround of course only, but it also explains why this is not seen in masses. I will of course want to be compatible, and I can see how the call to `parse_args` is something to make if it doesn't break with CPython.

**Comment by brianyuastri** [comment]
perhaps considering using `args = parser.parse_args('')` instead?

**Comment by brianyuastri** [comment]
`args = parser.parse_args('')` fixes the problem for me

**Comment by kayhayen** [comment]
I got this to work unchanged with what will be in 4.0, basically I am now using "pre load" module code for the "fake" module `__parents_main__` and that seems sufficient to solve it. It adapts the sys.argv to the compatible value, and only later restores it when `spawn_main` from the `multiprocessing` is called. That makes it work. I need to see that CI still likes the other test cases, but I guess that would have to be the case.

**Comment by kayhayen** [comment]
This is part of the stable release 4.0 that I just made.
